### PR TITLE
CLI commands for sign, verify

### PR DESCRIPTION
**Pull request by patrickxb**
These should be functionally equivalent to pgp sign, verify (same flags, etc.)

r? @maxtaco @oconnor663

### CODE REVIEW

File: go/client/cmd_sign.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	msg := ctx.String("message")
+	outfile := ctx.String("outfile")
+	infile := ctx.String("infile")
+
+	return s.FilterInit(msg, infile, outfile)
+}
+
+func (s *CmdSign) Run() (err error) {
+	protocols := []rpc.Protocol{
+		NewStreamUIProtocol(),
```
Comment (maxtaco) [inline go/client/cmd_sign.go]: i think it's time to contextify this protocol...

File: go/client/cmd_sign.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		NewSecretUIProtocol(s.G()),
+	}
+
+	cli, err := GetSaltPackClient(s.G())
+	if err != nil {
+		return err
+	}
+	if err = RegisterProtocols(protocols); err != nil {
```
Comment (maxtaco) [inline go/client/cmd_sign.go]: use the new `RegisterProtocolsWithContext`

File: go/client/cmd_verify.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			},
+			cli.StringFlag{
+				Name:  "m, message",
+				Usage: "Provide the message to verify on the command line.",
+			},
+			cli.StringFlag{
+				Name:  "S, signed-by",
+				Usage: "Assert signed by the given user (can use user assertion format).",
+			},
+		},
```
Comment (maxtaco) [inline go/client/cmd_verify.go]: have  `-o, --outfile` option.  And also a `--no-output` flag to keep the current behavior.

File: go/client/cmd_verify.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+
+	c.spui = &SaltPackUI{
+		Contextified: libkb.NewContextified(c.G()),
+		terminal:     c.G().UI.GetTerminalUI(),
+	}
+
+	protocols := []rpc.Protocol{
+		NewStreamUIProtocol(),
```
Comment (maxtaco) [inline go/client/cmd_verify.go]: contextify here...

File: go/client/cmd_verify.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		NewSecretUIProtocol(c.G()),
+		NewIdentifyUIProtocol(c.G()),
+		keybase1.SaltPackUiProtocol(c.spui),
+	}
+
+	if err = RegisterProtocols(protocols); err != nil {
```
Comment (maxtaco) [inline go/client/cmd_verify.go]: use `RegisterProtocolsWithContext`

File: go/client/cmd_verify.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	c.spui = &SaltPackUI{
+		Contextified: libkb.NewContextified(c.G()),
+		terminal:     c.G().UI.GetTerminalUI(),
+	}
+
+	protocols := []rpc.Protocol{
+		NewStreamUIProtocol(),
+		NewSecretUIProtocol(c.G()),
+		NewIdentifyUIProtocol(c.G()),
+		keybase1.SaltPackUiProtocol(c.spui),
```
Comment (maxtaco) [inline go/client/cmd_verify.go]: is this new style ok?  I eliminated a lot of cruft to do it directly.

File: go/client/saltpack_ui.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+func (s *SaltPackUI) SaltPackSignatureSuccess(_ context.Context, arg keybase1.SaltPackSignatureSuccessArg) error {
+	var un string
+	if arg.Sender.SenderType == keybase1.SaltPackSenderType_UNKNOWN {
+		un = "The signer of this message is unknown to Keybase"
+	} else {
+		un = fmt.Sprintf("Signed by %s", ColorString("bold", arg.Sender.Username))
+	}
+	s.terminal.Printf(ColorString("green", fmt.Sprintf("Signature verified. %s.\n", un)))
+	if arg.Sender.SenderType == keybase1.SaltPackSenderType_UNKNOWN {
+		s.terminal.Printf(ColorString("green", fmt.Sprintf("Signing key ID: %x.\n", arg.SigningKID.ToShortIDString())))
```
Comment (maxtaco) [inline go/client/saltpack_ui.go]: I'd say do the full KID here.

File: go/client/saltpack_ui.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func (s *SaltPackUI) SaltPackSignatureSuccess(_ context.Context, arg keybase1.SaltPackSignatureSuccessArg) error {
+	var un string
+	if arg.Sender.SenderType == keybase1.SaltPackSenderType_UNKNOWN {
+		un = "The signer of this message is unknown to Keybase"
+	} else {
+		un = fmt.Sprintf("Signed by %s", ColorString("bold", arg.Sender.Username))
+	}
+	s.terminal.Printf(ColorString("green", fmt.Sprintf("Signature verified. %s.\n", un)))
```
Comment (maxtaco) [inline go/client/saltpack_ui.go]: make sure this gets sent to stderr so it doesn't conflict with stdout in the case of a sig verification outputting payload.

File: go/client/cmd_verify.go (hunk 6 above)
Comment (patrickxb) [inline go/client/cmd_verify.go]: it's fine with me.  less code, right?

File: go/engine/saltpack_sign.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+		Me:      me,
+		KeyType: libkb.DeviceSigningKeyType,
+	}
+	key, err := e.G().Keyrings.GetSecretKeyWithPrompt(ctx.LoginContext, ska, ctx.SecretUI, "signing a message/file")
+	if err != nil {
+		return err
+	}
+	kp, ok := key.(libkb.NaclSigningKeyPair)
+	if !ok || kp.Private == nil {
+		return libkb.KeyCannotDecryptError{}
```
Comment (maxtaco) [inline go/engine/saltpack_sign.go]: need a more suitable error here....

File: go/libkb/interfaces.go [hunk 10]
```diff
@@ -302,6 +302,7 @@ type SecretUI interface {
 
 type SaltPackUI interface {
 	SaltPackPromptForDecrypt(context.Context, keybase1.SaltPackPromptForDecryptArg) error
+	SaltPackSignatureSuccess(context.Context, keybase1.SaltPackSignatureSuccessArg) error
```
Comment (maxtaco) [inline go/libkb/interfaces.go]: `SaltPackVerifySuccess` might be a better name?

File: protocol/avdl/saltpack_ui.avdl [hunk 11]
```diff
@@ -21,4 +21,5 @@ protocol saltPackUi {
   // and if something fishy might have happened.
   void saltPackPromptForDecrypt(int sessionID, SaltPackSender sender);
 
+  void saltPackSignatureSuccess(int sessionID, KID signingKID, SaltPackSender sender);
```
Comment (maxtaco) [inline protocol/avdl/saltpack_ui.avdl]: see above about a better name for this RPC

File: go/engine/saltpack_sender_identify.go [hunk 12]
```diff
@@ -88,6 +89,7 @@ func (e *SaltPackSenderIdentify) identifySender(ctx *Context) (err error) {
 		ForceRemoteCheck:      e.arg.forceRemoteCheck,
 		NoErrorOnTrackFailure: true,
 		Reason:                e.arg.reason,
+		UserAssertion:         e.arg.userAssertion,
```
Comment (maxtaco) [inline go/engine/saltpack_sender_identify.go]: this is OK for now, but long-term, we should try a strategy that supports OR-assertions. with a rebase, it shouldn't be too bad.

File: go/engine/saltpack_verify.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	return libkb.SaltPackVerifyDetached(e.G(), e.arg.Source, e.arg.Opts.Signature, hook)
+}
+
+func (e *SaltPackVerify) identifySender(ctx *Context, key saltpack.SigningPublicKey) (err error) {
+	defer e.G().Trace("SaltPackVerify::identifySender", func() error { return err })()
+
+	kid := libkb.SigningPublicKeyToKeybaseKID(key)
+	spsiArg := SaltPackSenderIdentifyArg{
+		publicKey:   kid,
+		interactive: true,
```
Comment (maxtaco) [inline go/engine/saltpack_verify.go]: this should be `false` i think

**Comment by maxtaco** [comment]
Looking good, let me play around with it some more on the command line.

**Comment by patrickxb** [comment]
Everything should be addressed now, so please take another look.

File: go/libkb/saltpack_sign.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	if _, err := io.Copy(stream, source); err != nil {
+		return err
+	}
+
+	if err = stream.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (oconnor663) [inline go/libkb/saltpack_sign.go]: I'm not very familiar with our WriteCloser idioms, but this function feels pretty generic. Is there some not-saltpack-specific adapter we could use here?

**Comment by oconnor663** [comment]
Just the one comment. Looks good to me.

The main issue I keep thinking about (not directly related to the code in this diff) is what happens when we display the sender but then the message doesn't actually authenticate. Anyone can create a message that _claims_ I'm the sender, and a fake message like that would still cause an `id` display for me. Maybe something to explain in our error messages?

**Comment by maxtaco** [comment]
Well at least we still have the constraint that no unsigned data is output, right?

**Comment by maxtaco** [comment]
LGTM, though we have a big performance problem. Likely in all saltpacks.

**Comment by oconnor663** [comment]
Yep, no unauthenticated data.

**Comment by patrickxb** [comment]
@oconnor663 we could move the checkSender to the end of the stream...

**Comment by oconnor663** [comment]
The plan was to prevent you from getting any data before you approved of who it came from, right? I guess we can't have our cake and eat it too, unless we want to use some giant temp files. Probably not worth changing anything.

**Pull request merged**